### PR TITLE
187 support flow types

### LINKED ISSUE #187: Support Flow typing in component_loader
## Description
Right now, creating a custom Dash component requires using the `prop-types` library. `react-docgen` used in `dash-components-archetype` [here](https://github.com/plotly/dash-components-archetype/blob/master/package.json#L37) does support Flow types, but the `metadata.json` parser in `component_loader` uses the fields generated from the `PropTypes` attribute of the component. Development of new components would be a potentially better environment if it also supported using Flow (see, e.g., a [`material-ui` `Dialog` component](https://github.com/StratoDem/sd-material-ui/blob/master/src/components/SDDialog.react.js#L57). It would also allow faster porting of React components already written with Flow to Dash components. 

### Necessary pieces
- [ ] The [eslint file](https://github.com/plotly/dash-components-archetype/blob/master/config/eslint/eslintrc-react.json) in `dash-components-archetype` will need to use `eslint-plugin-flowtype`
- [ ] The [babelrc file](https://github.com/plotly/dash-components-archetype/blob/master/config/babel/babelrc) in `dash-components-archetype` will need to use [`transform-class-properties`](https://babeljs.io/docs/plugins/transform-class-properties/) in order to support `static defaultProps = {}` (not required, but would be nice), and the [`react` preset already handles stripping Flow types](https://www.npmjs.com/package/babel-preset-react). 
- [ ] `package.json` in `dash-components-archetype` would need (at least) `eslint-plugin-flowtype` and `babel-plugin-transform-class-properties`
- [ ] One of the following two to address the metadata: 
    - [ ] `metadata.json` will need to be generated using the generated Flow types (matching the current structure) in `dash-components-archetype`, if they exist or 
    - [ ] `component_loader` will need to support Flow types if they are found in the `metadata.json` and the fields generated by `PropTypes` are not found.

Note that some of the code changes would occur in `dash-components-archetype`, but figured here might be a better place to discuss since the `component_loader` is the biggest "Dash user"-facing piece. There may also be other necessary pieces that I missed. 

We've set this structure up in the [`material-ui` ports](https://github.com/StratoDem/sd-material-ui) that we're using internally for our own material Dash components, and very much like it (we're also using jest for testing and would *love* to get the Dash toolchain updated to use [React 16](https://github.com/plotly/dash-components-archetype/issues/35) since the current chain is tied to an older version of React).

If this makes sense, I can create issues specifically on `dash-components-archetype` and plan to get this out in the next couple of weeks. There should not be any breaking changes anywhere from addressing this issue. 

Finally: *insert flamewar about type hinting here* :smile:

### PR DESCRIPTION
**Pull request by mjclawar**
### Description
Supports creating docstrings using `react-docgen` for Dash components written _solely_ with Flow type annotations. This is a non-breaking change.

The new logic, as determined in discussion in #187 is:
1. If a Dash component has `PropTypes`-generated typing, the docstring uses the `PropTypes`, _regardless of whether the component also has Flow types (current behavior)._
2. Otherwise if a Dash component has Flow types but _not `PropTypes`_, the docstring now uses the objects generated by `react-docgen` from the Flow types (**new behavior**).
 
### Implementation
There are no breaking changes, and all pre-existing tests for the `base_component` suite still pass. Unit tests were added by @coralvanda for the docstrings generated by Flow. The unit tests only added another `metadata.json` file for a Flow-generated `metadata.json` test example, and a [`TestFlowMetaDataConversions` `TestCase`](https://github.com/plotly/dash/blob/20f4b042948282d93dee559084c1c383bf5a65d7/tests/development/test_base_component.py#L760). 

As an example, the new Flow support on the Python `dict` in [this gist](https://gist.github.com/mjclawar/f108ca3d5d3f9ad55016816e6edf202b) will create this docstring:
```python
"""
A MyComponent component.
test description

Keyword arguments:
- requiredUnion (string | number; required)
- optionalString (string; optional): A string that isn't required.
- optionalBoolean (boolean; optional): A boolean test
- optionalArray (list; optional): An array test with a particularly 
long description that covers several lines. It includes the newline character 
and should span 3 lines in total.
- requiredString (string; required): A required string
- optionalSignature(shape) (optional): This is a test of an object's shape. optionalSignature(shape) has the following type: dict containing keys 'checked', 'children', 'customData', 'disabled', 'label', 'primaryText', 'secondaryText', 'style', 'value'.
  Those keys have the following types: 
  - checked (boolean; optional)
  - children (a list of or a singular dash component, string or number; optional)
  - customData (bool | number | str | dict | list; required): A test description
  - disabled (boolean; optional)
  - label (string; optional)
  - primaryText (string; required): Another test description
  - secondaryText (string; optional)
  - style (dict; optional)
  - value (bool | number | str | dict | list; required)
- requiredNested (required): . requiredNested has the following type: dict containing keys 'customData', 'value'.
  Those keys have the following types: 
  - customData (required): . customData has the following type: dict containing keys 'checked', 'children', 'customData', 'disabled', 'label', 'primaryText', 'secondaryText', 'style', 'value'.
    Those keys have the following types: 
    - checked (boolean; optional)
    - children (a list of or a singular dash component, string or number; optional)
    - customData (bool | number | str | dict | list; required)
    - disabled (boolean; optional)
    - label (string; optional)
    - primaryText (string; required)
    - secondaryText (string; optional)
    - style (dict; optional)
    - value (bool | number | str | dict | list; required)
  - value (bool | number | str | dict | list; required)
- optionalNode (a list of or a singular dash component, string or number; optional): A node test
"""
``` 

Thanks to @coralvanda, this now also supports nested object definitions for Flow, e.g. this section:
```
- requiredNested (required): . requiredNested has the following type: dict containing keys 'customData', 'value'.
  Those keys have the following types: 
  - customData (required): . customData has the following type: dict containing keys 'checked', 'children', 'customData', 'disabled', 'label', 'primaryText', 'secondaryText', 'style', 'value'.
    Those keys have the following types: 
    - checked (boolean; optional)
    - children (a list of or a singular dash component, string or number; optional)
    - customData (bool | number | str | dict | list; required)
    - disabled (boolean; optional)
    - label (string; optional)
    - primaryText (string; required)
    - secondaryText (string; optional)
    - style (dict; optional)
    - value (bool | number | str | dict | list; required)
  - value (bool | number | str | dict | list; required)
```
which is pulled recursively from the `requiredNested` prop found [here in the gist](https://gist.github.com/mjclawar/f108ca3d5d3f9ad55016816e6edf202b#file-dash-metadata-with-flow-example-py-L265).

### Related issue
Closes #187 
Bumps version to 0.20.1, assuming this is appropriate.

### CODE REVIEW

**Comment by chriddyp** [comment]
very nice! will take a look at this this week. do you have an example of a Dash react component that uses Flow? am curious about what the user would need to modify after generating the boilerplate from the `dash-components-archetype` package.

**Comment by mjclawar** [comment]
I don't believe a user would need to modify anything if you use the current boilerplate. The babel react preset already supports stripping Flow types. You would `$ npm install flow-bin` as a developer and set it up in your build environment (we use IntelliJ, so it has linting for it).

[Here's a `FlatButton` component](https://github.com/StratoDem/sd-material-ui/blob/v2.0.0/src/components/FlatButton/FlatButton.react.js) that currently has both Flow and `PropTypes`. With this PR, we would just delete the `PropTypes` and move the descriptions from `PropTypes` to above the current Flow type prop and everything would work the same. The `sd-material-ui` package would then have a breaking change for the next release, since we would not support a dash version for `metadata.json` parsing below a potential release via this PR.

As a side note, pylint is very aggressive and not playing nice with my PEP8 linting. I think I finally got it where the tests pass.

**Comment by mjclawar** [comment]
@chriddyp Need anything else on this? Also ran the codemods to help out with optionally updating the bundled React version for `dash-html-components` and `dash-core-components` this weekend.

File: dash/development/base_component.py [hunk 1]
```diff
@@ -266,159 +283,372 @@ def __repr__(self):
...
-                )
-            )
-        )
-    }
+        shape=lambda: 'dict containing keys {}.\n{}'.format(
+            ', '.join(
+                "'{}'".format(t)
+                for t in list(type_object['value'].keys())),
+            'Those keys have the following types: \n{}'.format(
+                '\n'.join(create_prop_docstring(
```
Comment (chriddyp) [inline dash/development/base_component.py]: 🙇 🙇 🙇 🙇

**Comment by chriddyp** [comment]
@mjclawar and @coralvanda - Sorry for the delay on this one. I've finally found some time to review and this looks good to me. Many, many thanks for writing the thorough tests and taking the time to improve the existing code.

Is there anything else that you would like to add? Otherwise, I'll merge this tonight and make a new release.

I think that I'd like to make a separate repo containing a sample flowtypes component that I can include across Dash's end-to-end integration tests. I can do this separately and it shouldn't block this PR.

**Comment by chriddyp** [comment]
> I think that I'd like to make a separate repo containing a sample flowtypes component that I can include across Dash's end-to-end integration tests

I created this here: https://github.com/plotly/dash-flow-example. I'll add an integration test that includes these prop types once this PR is merged. Screenshot of the example app:
image

@mjclawar or @coralvanda - How would I convert these nested `PropTypes` into flow types?
```

    /**
     * A list of options
     * Multiline description
     */
    options: PropTypes.arrayOf(PropTypes.shape({
        /**
         * The label of the option
         */
        'label': PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

        /**
         * The value of the option
         */
        'value': PropTypes.string,
    })),

    /**
     * A config object
     */
    config: PropTypes.shape({
        /**
         * The title option
         */
        'title': PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

        /**
         * The value of the option
         */
        'value': PropTypes.string,
    }),
```
I'd like to add a couple of nested properties to the flow example: https://github.com/plotly/dash-flow-example/blob/master/src/components/ExampleFlowComponent.react.js#L3

**Comment by mjclawar** [comment]
@chriddyp nice!

```
 /**
     * A list of options
     * Multiline description
     */
    options: PropTypes.arrayOf(PropTypes.shape({
        /**
         * The label of the option
         */
        'label': PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

        /**
         * The value of the option
         */
        'value': PropTypes.string,
    })),

    /**
     * A config object
     */
    config: PropTypes.shape({
        /**
         * The title option
         */
        'title': PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

        /**
         * The value of the option
         */
        'value': PropTypes.string,
    }),
```

is equivalent to
```js
{
  /* A list of options*/
  options: Array<{
    /* The label of the option */
    label: string | number, 
    /* The value of the option*/
    value: string,
  }>,
  /* A config object */
  config: {
    /* The title option*/
    title: string | number,
    /* The value of the option */
    value: string,
}
```

You also can create a temporary type object to use elsewhere throughout the code, which is the best part! Example:
```js
type T_OPTION = {
  /* The label of the option */
  label: string | number, 
  /* The value of the option*/
  value: string,
};
type T_CONFIG = {
  /* The title option */
  title: string | number,
  /* The value of the option */
  value: string
};

...
{
  /* A list of options */
  options: Array<T_OPTION>,
  config: T_CONFIG,
}
```

Once this is updated, more than willing to pass along a couple of examples from our libraries for more Flow examples.

**Comment by mjclawar** [comment]
> Is there anything else that you would like to add? Otherwise, I'll merge this tonight and make a new release.

Nothing besides wishing there were also typescript support for completeness, but I have never used that and have no interest in learning tonight! I think this PR completely addresses the issue. Plenty of examples can follow in other repos, and I would guess that the https://plot.ly/dash/plugins page could benefit from some updates after this.

**Comment by chriddyp** [comment]
Great! Nice work @mjclawar  and @coralvanda  🍻

**Comment by chriddyp** [comment]
I'm going to wait until #212 passes tests before making the formal release.

**Comment by mjclawar** [comment]
:crossed_fingers:

**Comment by chriddyp** [comment]
so far so good! available in `v0.21.0`. Doing the final screenshot test sanity check with the dash userguide https://github.com/plotly/dash-docs/pull/69

**Comment by mjclawar** [comment]
@chriddyp see updated `sd-material-ui` package with examples of using Flow types.

e.g., [`Checkbox`](https://github.com/StratoDem/sd-material-ui/blob/v2.1.1/src/components/Checkbox/Checkbox.react.js#L10-L33)
```js
type Props = {
  /** Checkbox is checked if true */
  checked?: boolean,
  /** Checkbox is disabled if true */
  disabled?: boolean,
  /** A callback for firing events to dash */
  fireEvent?: () => void,
  /** Overrides the inline-styles of the icon element */
  iconStyle?: Object,
  /** The element's ID */
  id: string,
  /** Overrides the inline styles of the input element */
  inputStyle?: Object,
  /** The text label for the checkbox */
  label?: string,
  /** Where the label will be placed next to the checkbox */
  labelPosition?: 'left' | 'right',
  /** Overrides the inline styles of the Checkbox element label */
  labelStyle?: Object,
  /** Dash callback to update props on the server */
  setProps?: () => void,
  /** Override the inline styles of the root element */
  style?: Object,
};
```

and [`Snackbar`](https://github.com/StratoDem/sd-material-ui/blob/v2.1.1/src/components/Snackbar/Snackbar.react.js#L10-L44)
```js
type Props = {
  /** Label for the action on the snackbar */
  action?: Node,
  /**
   * The number of milliseconds to wait before automatically dismissing. If no value is specified
   * the snackbar will dismiss normally. If a value is provided the snackbar can still be dismissed
   * normally. If a snackbar is dismissed before the timer expires, the timer will be cleared.
   */
  autoHideDuration?: number,
  /** Override the inline styles of the body element */
  bodyStyle?: object,
  /** CSS class name of the root element */
  className?: string,
  /** Override the inline styles of the content element */
  contentStyle?: object,
  /** Dash event handler for click events */
  fireEvent?: () => void,
  /** The element's ID */
  id: string,
  /**
   * The message to be displayed.
   * (Note: If the message is an element or array, and the Snackbar may re-render while it is
   * still open, ensure that the same object remains as the message property if you want to avoid
   * the Snackbar hiding and showing again)
   */
  message: Node,
  /** An integer that represents the number of times that action button has been clicked */
  n_clicks?: number,
  /** Controls whether the Snackbar is opened or not */
  open: boolean,
  /** Dash callback to update props on the server */
  setProps?: (props: {open?: boolean}) => void,
  /** Override the inline styles of the root element */
  style?: object,
};
```

**Pull request merged**